subprojects/docs/src/docs/userguide/publishingOverview.adoc

```diff
@@ -0,0 +1,262 @@
+
+The following sample demonstrates how you can do this by defining an aggregate task — `publishToExternalRepository` — that depends on all the relevant publish tasks:
+
+++++
+<sample dir="maven-publish/conditional-publishing" id="shorthandTasks" title="Defining your own shorthand tasks for publishing">
+    <sourcefile file="build.gradle" snippet="shorthand-tasks"/>
+</sample>
+++++
+
+This particular sample automatically handles the introduction or removal of the relevant publishing tasks by using api:org.gradle.api.tasks.TaskCollection#withType(java.lang.Class)[] with the api:org.gradle.api.publish.maven.tasks.PublishToMavenRepository[] task type. You can do the same with api:org.gradle.api.publish.ivy.tasks.PublishToIvyRepository[] if you're publishing to Ivy-compatible repositories.
+
+[[sec:configuring_publishing_tasks]]
+=== Configuring publishing tasks
+
+The publishing plugins create their non-aggregate tasks after the project has been evaluated, which means you cannot directly reference them from your build script. If you would like to configure any of these tasks, you should use deferred task configuration. This can be done in a number of ways via the project's `tasks` collection.
+
+For example, imagine you want to change where the `generatePomFileFor__PubName__Publication` tasks write their POM files. You can do this by using the api:org.gradle.api.tasks.TaskCollection#withType(java.lang.Class)[] method, as demonstrated by this sample:
+
+++++
+<sample dir="userguide/publishing/javaLibrary" id="deferred-configuration-publishing-task" title="Configuring a dynamically named task created by the publishing plugins">
+    <sourcefile file="build.gradle" snippet="configure-generate-task"/>
+</sample>
+++++
+
+The above sample uses a regular expression to extract the name of the publication from the name of the task. This is so that there is no conflict between the file paths of all the POM files that might be generated. If you only have one publication, then you don't have to worry about such conflicts since there will only be one POM file.
+
+=== Terminology
+
+[[glossary:artifact]]
+Artifact::
+A file produced by a build, such as a JAR, a ZIP distribution, or a native executable.
++
+Artifacts are typically designed to be used or consumed by users or other projects, or deployed to hosting systems. This is distinct from intermediate files that are created by a build as a step on the way to producing the end result. For example, the class files generated by compiling Java source are almost never an end product themselves, since they are typically packaged into JARs or some other type of archive.
+
+[[glossary:component]]
+Component::
+A predefined combination of one or more artifacts and a standard set of information about those artifacts. The concept exists in Gradle only as a part of publishing.
```

Comment:
The difference are the dependencies at the moment. For Maven we merge the two variants into one POM, because there is no other way. API dependencies go into `<compile>` scope, runtime dependencies go into `<runtime>` scope.

For Gradle metadata the variants are separated more explicitly. Both currently point to the same artifact, because we don't create an API jar (we have no DSL for defining what packages make up your API).

Reply:
There's still only one JAR, right? I was assuming that variants would have different artifacts, but you seem to be suggesting that's not the case.

Reply:
Not *yet*. We could certainly publish a different API jar in the future. Not for the POM, but for Gradle metadata, which is more accurate and flexible. The POM would keep referencing the full jar.